backend/src/cache/server/mutation.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		log.Printf("Unable to generate cache key for pod %s : %s", pod.ObjectMeta.Name, err.Error())
+		return patches, nil
+	}
+
+	annotations[ExecutionKey] = executionHashKey
+	labels[CacheIDLabelKey] = ""
+
+	var cachedExecution *model.ExecutionCache
+	cachedExecution, err = clientMgr.CacheStore().GetExecutionCache(executionHashKey)
+	if err != nil {
+		log.Println(err.Error())
+	}
+	// Found cached execution, add cached output and cache_id and replace container images.
+	if cachedExecution != nil {
+		log.Println("Cached output: " + cachedExecution.ExecutionOutput)
+
+		annotations[ArgoWorkflowOutputs] = getOutputFromOutputMap(cachedExecution.ExecutionOutput)
+		labels[CacheIDLabelKey] = strconv.FormatInt(cachedExecution.ID, 10)
+		labels[MetadataExecutionIDKey] = getMetadataExecutionIDFromOutputMap(cachedExecution.ExecutionOutput)
+		labels[MetadataWritenKey] = "true"
+
+		dummyContainer := corev1.Container{
+			Name:    "main",
+			Image:   "alpine",
+			Command: []string{`echo`, `"This step output is taken from cache."`},
+		}
+		dummyContainers := []corev1.Container{
+			dummyContainer,
+		}
+		patches = append(patches, patchOperation{
+			Op:    OperationTypeReplace,
+			Path:  SpecContainersPath,
+			Value: dummyContainers,
+		})
+		if pod.Spec.InitContainers != nil || len(pod.Spec.InitContainers) != 0 {
+			patches = append(patches, patchOperation{
+				Op:   OperationTypeRemove,
+				Path: SpecInitContainersPath,
+			})
+		}
+	}
+
+	// Add executionKey to pod.metadata.annotations
+	patches = append(patches, patchOperation{
+		Op:    OperationTypeAdd,
+		Path:  AnnotationPath,
+		Value: annotations,
+	})
+
+	// Add cache_id label key
+	patches = append(patches, patchOperation{
+		Op:    OperationTypeAdd,
+		Path:  LabelPath,
+		Value: labels,
+	})
+
+	return patches, nil
+}
+
+func generateCacheKeyFromTemplate(template string) (string, error) {
+	var templateMap map[string]interface{}
+	b := []byte(template)
+	err := json.Unmarshal(b, &templateMap)
+	if err != nil {
+		return "", err
+	}
+
+	// template[archiveLocation] needs to be removed when calculating cache key.
+	// Because archiveLocation.key is different in every single run.
+	_, exists := templateMap[ArchiveLocationKey]
+	if exists {
+		log.Println("ArchiveLocation exists in template.")
+		delete(templateMap, ArchiveLocationKey)
+	}
+	b, err = json.Marshal(templateMap)
+	if err != nil {
+		return "", err
+	}
+	hash := sha256.New()
+	hash.Write(b)
+	md := hash.Sum(nil)
+	executionHashKey := hex.EncodeToString(md)
+
+	return executionHashKey, nil
+}
+
+func getOutputFromOutputMap(output string) string {
+	var outputMap map[string]interface{}
+	b := []byte(output)
+	err := json.Unmarshal(b, &outputMap)
+	if err != nil {
+		return ""
+	}
+
+	executionOutput, exist := outputMap[ArgoWorkflowOutputs].(string)
+	if executionOutput == "" || !exist {
+		return ""
+	}
+	return executionOutput
+}
+
+func getMetadataExecutionIDFromOutputMap(output string) string {
```

Comment:
This function looks similar to `getOutputFromOutputMap`. Should we just have a single `func getValueFromSerializedMap(key string) string`?